Enforce western arabic numerals for Stats

ovitrif opened the pull request:
Fixes Partially #17002 — Handles only the Stats screens and the Today Stats card on My Site Dashboard

This PR introduces a wrapper over `MaterialTextView` that enforces the use of Western Arabic numerals by replacing Eastern Arabic and Persian/Urdu numerals to their Western Arabic counterparts.

This UI component was then applied to every stats widget which uses `MaterialTextView` and displays numerals.

The purpose of this PR is to make sure we're always displaying Western Arabic numerals on the Stats screens, to mainly address the remark from the issue:
> _This looks especially strange on screens where we use both types of numerals_

To test:

##### 1️⃣ Test Case 1 : WordPress App Stats

1. Open the WordPress App, Go to `Me` → `App Settings` → `Interface Language` and select `Arabic`.
2. Go back to `My Site` → `Home` tab
   - Expect the numbers on `Today's Stats` card to be western arabic.
3. Go to `Stats` & explore the first tab (Insights)
   - Expect all displayed numbers to be western arabic.
4. Go to any other `Stats` tab (Days/Weeks/...)
   - Expect all displayed numbers to be western arabic.

##### 1️⃣ Test Case 2 : Jetpack App Stats

1. Open the Jetpack App, Go to `Me` → `App Settings` → `Interface Language` and select `Arabic`.
2. Repeat steps 2-4 from Test Case 1
3. Tap "View More" on one of the stats cards having a different value than `0`
   - Expect numbers to be Western Arabic
   - Expect the numeral inside the Pie Chart to be of larger font size than the word `Totals`

<details>
<summary>Preview Test Case 2 — Step 3</summary>

<img width="314" src="https://user-images.githubusercontent.com/4588074/190453085-8e028195-a23e-450f-a379-770ba2997ede.png">

**Note** For easier understanding the screenshot is in English. Same result should be expected in Arabic.

---

</details>

### Previews

#### Stats → Insights tab
| Before | After |
| --- | --- |
| ![stats_insights](https://user-images.githubusercontent.com/4588074/190380852-ef802513-ecaa-4b67-8304-9566630dc24b.png) | ![stats_followers_fix](https://user-images.githubusercontent.com/4588074/190380863-99f51981-ce97-4d7b-a7ca-dfa7c0dc5fe4.png) |

## Regression Notes
1. Potential unintended areas of impact
   Unexpected, since we're only replacing Indo-Arabic and Persian/Urdu numerals with the normal (Western) Arabic numerals.

6. What I did to test those areas of impact (or what existing automated tests I relied on)
   Tested manually in Chinese & English languages.

7. What automated tests I added (or what prevented me from doing so)
   N/a, we're only touching the view layer with these changes.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ovitrif committed:
Implement MaterialTextViewWithArabicNumerals

ovitrif committed:
Use MaterialTextViewWithArabicNumerals for My Site Todays Stats card

ovitrif committed:
Use MaterialTextViewWithArabicNumerals for Stats widgets

ovitrif committed:
Move enforceWesternArabicNumerals function to class scope and make it private

ovitrif committed:
Update release notes for Western Arabic numerals fix

antonis commented on WordPress/src/main/java/org/wordpress/android/widgets/MaterialTextViewWithArabicNumerals.kt:
```diff
@@ -0,0 +1,24 @@
+/**
+ * MaterialTextView which enforces Western Arabic numerals.
+ */
+class MaterialTextViewWithArabicNumerals @JvmOverloads constructor(
```
I like the approach of wrapping the `MaterialTextView` as it provides a cleaner solution 👍 
I'm a bit undecided about the naming `MaterialTextViewWithArabicNumerals` though. I feel that a more generic name (e.g. `MaterialTextViewWithNumerals`) might be more future proof in case we need to replace something else in the future.

ovitrif replied:
Good idea, I've renamed the view to `MaterialTextViewWithNumerals` in https://github.com/wordpress-mobile/WordPress-Android/pull/17160/commits/6eb6b4b8bdbef713ff6a7934e95922d8814b14f5. Thank you @antonis 🙇

antonis reviewed (changes requested):
Great work @ovitrif 👍 
The code LGTM and the app behaves as expected in most cases. I've noticed an issue with zero not being transformed.
|Arabic|English|
|---|---|
|![Screenshot_20220915_161555](https://user-images.githubusercontent.com/304044/190415239-2a4d35ab-a864-438f-9b63-1de4923af90d.png)|![Screenshot_20220915_161609](https://user-images.githubusercontent.com/304044/190415245-e852bb48-981f-4910-9123-76fde690418d.png)|

I've also left a comment which is more of a suggestion 🙏

ovitrif committed:
Rename MaterialTextViewWithArabicNumerals into MaterialTextViewWithNumerals

ovitrif commented:
> Great work @ovitrif 👍 The code LGTM and the app behaves as expected in most cases. I've noticed an issue with zero not being transformed.

I'm trying to figure out how to add the views from your screenshot to my Stats. On the emulator that I'm testing (Pixel 5, Api 28) the issue doesn't seem to be specific to `0`, which makes me guess if there's some different views where I didn't replace `MaterialTextView` with `MaterialTextViewWithNumerals` 🤷 

I added a few more stats cards but couldn't yet replicate the ones in your screenshot:

| English | Arabic |
| --- | --- |
| ![stats_en](https://user-images.githubusercontent.com/4588074/190428374-6d715d51-3e56-404f-93a2-b86e8f620de2.png) | ![stats_ar](https://user-images.githubusercontent.com/4588074/190428172-62af9e47-5849-41b3-867d-7498431e155a.png) |

I'll try with the Jetpack App 👍

ovitrif commented:
> I'll try with the Jetpack App 👍

I was able to reproduce the issue with the Jetpack App. Working on a fix 👍

antonis commented:
Thank you for iterating on this @ovitrif 👍 

> On the emulator that I'm testing (Pixel 5, Api 28) the issue doesn't seem to be specific to 0

Sorry for not providing more info on my testing setup. I've tested on a Pixel 5 with Android 13 (API level 33). I wonder if this can be Android version specific.

> which makes me guess if there's some different views where I didn't replace MaterialTextView with MaterialTextViewWithNumerals 🤷

That might be the case since I was replicating this on the first/insights screen. The other screens LGTM

https://user-images.githubusercontent.com/304044/190430517-dc30cde9-1baf-493a-811c-ad8e9fd888c9.mp4

ovitrif committed:
Use MaterialTextViewWithNumerals in JP stats widgets

ovitrif committed:
Preserve spans in MaterialTextViewWithNumerals

ovitrif commented:
@antonis Thank you for testing this with the Jetpack app. There were indeed different views where we had to use the `MaterialTextViewWithNumerals` instead of the `MaterialTextView`.

I've updated all views I could find on the Jetpack App

| JP Stats - Top | JP  Stats - Bottom |
| --- | --- |
| ![stats_jp_top](https://user-images.githubusercontent.com/4588074/190450999-05629a14-f1b3-48e2-94ee-15cecd21a4d5.png) | ![stats_jp_bottom](https://user-images.githubusercontent.com/4588074/190451003-a5c16034-9276-4909-83a4-3fa41c11a73b.png) |

There was a special case I had to fix, this appears when going to the details of a stats card which has a different value than zero (see screenshot):

<img width="314" src="https://user-images.githubusercontent.com/4588074/190451567-ae9b62bd-e2ee-4a9b-b2d7-5789a11adcff.png">

In the view that appears when tapping the highlighted button, the text in the middle of the pie chart was decorated with Spans (it's a `SpannableString`). I had to update `MaterialTextViewWithNumerals` to preserve the spans.

Previews:

| Before preserving Spans | After |
| --- | --- |
| ![stats_jp_details_fix](https://user-images.githubusercontent.com/4588074/190452133-cc514d7b-b143-4991-a66f-ecebea81a552.png) | ![stats_jp_details_fix_spans](https://user-images.githubusercontent.com/4588074/190452159-36550b2e-350b-4184-9da6-8d2ac96361bf.png) |

---

I updated the testing steps and added a test case for the Jetpack App 🙇

antonis reviewed (approved):
Great work @ovitrif 👍 
I've retested the App CI builds with the latest changes and everything worked as expected on my Pixel 5 (Android 13). The code changes also LGTM 🎉 
Thank you for tackling this 🙇

ovitrif commented:
> Great work @ovitrif 👍 I've retested the App CI builds with the latest changes and everything worked as expected on my Pixel 5 (Android 13). The code changes also LGTM 🎉 Thank you for tackling this 🙇

Awesome! Many thanks for your help on this PR Antonis 🙇 

I'll fix the conflicts in `RELEASE-NOTES.txt` then we can merge the enhancement 🚀 

P.S. I've updated the milestone to `20.8` since we're going to make it in time 🐇

ovitrif committed:
Merge branch 'trunk' into issue/17002-use-western-arabic-numerals

ovitrif committed:
Update release notes
